hnn_core/parallel_backends.py

```diff
@@ -345,105 +498,11 @@ def simulate(self, net):
-        os.write(pipe_stdin_w, pickled_params)
-
-        # signal that we are done writing params
-        os.close(pipe_stdin_w)
-        os.close(pipe_stdin_r)
-
-        # create the selector instance and register all input events
-        # with self.read_stdout which will only echo to stdout
-        self.sel = selectors.DefaultSelector()
-        self.sel.register(pipe_stdout_r, selectors.EVENT_READ,
-                          self._read_stdout)
-        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
-                          self._read_stdout)
-
-        data_len = 0
-        completed = False
-        timeout = 0
-        # loop while the process is running
-        while True:
-            if not proc.poll() is None:
-                if completed is True:
-                    break
-                elif timeout > 4:
-                    # This is indicative of a failure. For debugging purposes.
-                    warn("Timed out (5s) waiting for end of data after child "
-                         "process stopped")
-                    break
-                else:
-                    timeout += 1
-                    sleep(1)
-
-            # wait for an event on the selector, timeout after 1s
-            events = self.sel.select(timeout=1)
-            for key, mask in events:
-                callback = key.data
-                completion_signal = callback(key.fileobj, mask)
-                if completion_signal is not None:
-                    if completion_signal == "end_of_sim":
-                        # finishied receiving printable output
-                        # everything else received is data
-                        self.sel.unregister(pipe_stderr_r)
-                        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
-                                          self._read_stderr)
-                    elif isinstance(completion_signal, int):
-                        data_len = completion_signal
-                        self.sel.unregister(pipe_stdout_r)
-                        completed = True
-
-                        # there could still be data in stderr, so we return
-                        # to waiting until the process ends
-                    else:
-                        raise ValueError("Unrecognized signal received from "
-                                         "MPI child")
-
-        # cleanup the selector
-        self.sel.unregister(pipe_stderr_r)
-        self.sel.close()
-
-        # done with stdout and stderr
-        os.close(pipe_stdout_r)
-        os.close(pipe_stdout_w)
-        os.close(pipe_stderr_r)
-        os.close(pipe_stderr_w)
-
-        # if simulation failed, raise exception
-        if proc.returncode != 0:
-            raise RuntimeError("MPI simulation failed")
-
-        sim_data = self._process_child_data(self.proc_data_bytes, data_len)
+        proc, sim_data = run_subprocess(
+            command=self.mpi_cmd, obj=net, timeout=4,
```

Comment:
Curious where this timeout value came from. Clearly it works currently, but I wonder if this can fail if we add more `CellResponse` recordings in the future.

Reply:
It's not related to how much data is being passed. That is handled by the "end_of_data" signal. This just waits long enough to catch error messages if the process still has some in its stderr buffer. Also, it's unlikely that there are any errors at this point, during the shutdown process, but this just covers that case.